polkadot/xcm/xcm-builder/src/nonfungible_adapter.rs

```diff
@@ -0,0 +1,319 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Adapters to work with [`frame_support::traits::tokens::nonfungible`] through XCM.
+
+use super::MintLocation;
+use frame_support::{
+	ensure,
+	traits::{tokens::nonfungible, Get},
+};
+use sp_std::{marker::PhantomData, prelude::*, result};
+use xcm::latest::prelude::*;
```

Comment:
XCM latest has updated to v4 in master since your PR, so some of your types need to have the `Multi` prefix removed and `xcm_executor::Assets` renamed to `xcm_executor::AssetsInHolding`. Updating from master and making those small changes are all you'll need to update your branch to v4. See https://github.com/paritytech/polkadot-sdk/blob/8428f678fe5179399f6854e194f1c8b30a9102f9/polkadot/xcm/xcm-builder/src/nonfungibles_adapter.rs as an example.

I've suggested the necessary changes in this review.